homeassistant/components/google.py

```diff
@@ -0,0 +1,294 @@
+            hass.services.call(DOMAIN, SERVICE_FOUND_CALENDARS,
+                               calendar)
+
+    hass.services.register(
+        DOMAIN, SERVICE_SCAN_CALENDARS,
+        _scan_for_calendars,
+        None, schema=None)
+    return True
+
+
+def do_setup(hass, config):
+    """Run the setup after we have everything configured."""
+    # load calendars the user has configured
+    hass.data[DATA_INDEX] = load_config(hass.config.path(YAML_DEVICES))
+
+    calendar_service = GoogleCalendarService(hass.config.path(TOKEN_FILE))
+    track_new_found_calendars = convert(config.get(CONF_TRACK_NEW),
+                                        bool, DEFAULT_CONF_TRACK_NEW)
+    setup_services(hass, track_new_found_calendars, calendar_service)
+
+    # Ensure component is loaded
+    bootstrap.setup_component(hass, 'calendar', config)
+
+    for calendar in hass.data[DATA_INDEX].values():
+        discovery.load_platform(hass, 'calendar', DOMAIN, calendar)
+
+    # look for any new calendars
+    hass.services.call(DOMAIN, SERVICE_SCAN_CALENDARS, None)
+    return True
+
+
+class GoogleCalendarService(object):
+    """Calendar service interface to google."""
+
+    def __init__(self, token_file):
+        """We just need the token_file."""
+        self.token_file = token_file
+
+    def get(self):
+        """Get the calendar service from the storage file token."""
+        import httplib2
+        from oauth2client.file import Storage
+        from googleapiclient import discovery as google_discovery
+        credentials = Storage(self.token_file).get()
+        http = credentials.authorize(httplib2.Http())
+        service = google_discovery.build('calendar', 'v3', http=http)
+        return service
+
+
+def get_calendar_info(hass, calendar):
+    """Convert data from Google into DEVICE_SCHEMA."""
+    calendar_info = DEVICE_SCHEMA({
+        CONF_CAL_ID: calendar['id'],
+        CONF_ENTITIES: [{
+            CONF_TRACK: calendar['track'],
+            CONF_NAME: calendar['summary'],
+            CONF_DEVICE_ID: generate_entity_id('{}', calendar['summary'],
+                                               hass=hass),
+        }]
+    })
+    return calendar_info
+
+
+def load_config(path):
+    """Load the google_calendar_devices.yaml."""
+    calendars = {}
+    try:
+        with open(path) as file:
+            data = yaml.load(file)
+            for calendar in data:
+                try:
+                    calendars.update({calendar[CONF_CAL_ID]:
+                                      DEVICE_SCHEMA(calendar)})
+                except VoluptuousError as exception:
+                    # keep going
+                    _LOGGER.warning('Calendar Invalid Data: %s', exception)
+    except FileNotFoundError:
+        # When YAML file could not be loaded/did not contain a dict
+        return {}
+
+    return calendars
+
+
+def update_config(path, calendar):
+    """Write the google_calendar_devices.yaml."""
+    with threading.Lock():
```

Comment:
This will create a new lock every time you call this, so that means that the lock is never shared -> the locking won't work